### DISCUSSION TITLE
Mockito 5.3.1 High Memory Usage

### DISCUSSION THREAD

**Discussion by debraj-manna** [mockito/mockito#3300] [Q&A]
With mock 5.3.1 I am observing a high usage of heap memory. Looking at the heap dump it looks like most of the memory are being taken by a Map containing `MockMethodInterceptor`.

<img width="1596" alt="Screenshot 2024-03-15 at 17 58 53" src="https://github.com/mockito/mockito/assets/8141860/a55aa417-7f4d-41df-9257-21a973e2ccd8">

I am not that familiar with the internal implementation of Mockito. Can someone please suggest what could be going wrong here or are there any known issue in the version of Mockito I am using?

**Comment by TimvdLippe** [comment]
Does this happen since 5.3.1 or does it happen before? Can you also test out several ByteBuddy versions to see if there is a regression with ByteBuddy instead?

**Reply by debraj-manna** [reply]
Looks like the byte-buddy that is getting used in our codebase is 1.12.23 from the dependency tree

```
org.mockito:mockito-core:jar:5.3.1:test
[INFO] |  |  \- net.bytebuddy:byte-buddy-agent:jar:1.12.23:test
```

Mockito has specified version 1.14.4 of byte-buddy. I will try using the byte-buddy of 1.14.4 or you suggest trying some other version?

**Comment by debraj-work** [comment]
@TimvdLippe I tried with version 1.14.4 and 1.14.14 version of byte-buddy. With both versions, I am observing the same behavior as 1.12.23. 

Is there a way for me to not use byte-buddy with mockito?

**Reply by TimvdLippe** [reply]
You can bisect if a previous version of Mockito has similar behavior. In case (for example) 4.9 does not show the same behavior, it means there is a regression in Mockito instead. If there is a difference in behavior, you can try to use `mockito-subclass`. If you don't want any of the pre-built mockmakers, you can implement the `MockMaker` interface yourself, but I want to warn you that it is tricky to implement correctly.

**Reply by debraj-work** [reply]
@TimvdLippe It looks like the same issue is reported in the quarkus project also. There using mockito-subclass solved the issue.

https://github.com/quarkusio/quarkus/issues/31251

In some places we are mocking static methods, if I am going to use mockito-subclass with mockito 5.3.1, how do I mock the static methods?

**Comment by debraj-manna** [comment]
In my case also adding `Mockito.framework().clearInlineMocks()` solved the issue as mentioned in https://github.com/quarkusio/quarkus/issues/31251

**Comment by jvdvegt** [comment]
We also experience high memory usage, with `org.mockito.internal.util.concurrent.WeakConcurrentMap$LatentKey` using more than 7GB  and `org.mockito.internal.invocation.InvocationsFinder$$Lambda+0x000...` using almost 1GB (as reported by enabling Java FlightRecorder and opening the report in Oracle's Mission Control)

This happens on OpenJDK 21.0.8 with Mockito 5.13.0 and 5.19.0.

As a work-around, I created a JUnit `BeforeAllCallback`-extension and have it invoke `Mockito.framework().clearInlineMocks()`. That reduced memory consumption by 1.5GB. Not as much as I hoped, but significant non the less.

For some tests, I got `org.mockito.exceptions.misusing.DisabledMockException: Mock accessed after inline mocks were cleared` errors. Those were:
- JUnit `@Nested` test classes - I had to remove those.
- Sprint `@WebAppConfiguration` test classes.
- A really complicated test class with a lot of `static` mocked field that were `reset(mock)` between tests.

I had to skip `clearInlineMocks` for the latter two situations.
